### PR TITLE
Add support for Binance testnet addresses.

### PR DESCRIPTION
**Pull request by andreibancioiu**
## Description

 - For Binance, added a new `Address` constructor, one that also receives the HRP as a parameter (a bit similar to the Cosmos implementation) - this allows one to derive testnet addresses, as well.
 - ~~Extended the array `validHrps` for Binance, to include `tbnb`, as well - **also adjusted a unit test for address validation, to take this change into consideration**.~~
 - Extend validation of `Binance/Address` - add a new function overload for address validation, one that handles a provided `hrp`.

Inspired by: https://github.com/trustwallet/wallet-core/pull/2643

## How to test

Run all unit tests.

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: swift/Tests/HDWalletTests.swift [hunk 1]
```diff
@@ -121,6 +121,9 @@ class HDWalletTests: XCTestCase {
 
         let address2 = wallet.getAddressDerivation(coin: coin, derivation: .bitcoinLegacy)
         XCTAssertEqual(address2, "1PeUvjuxyf31aJKX6kCXuaqxhmG78ZUdL1")
+
+        let address3 = wallet.getAddressDerivation(coin: coin, derivation: .bitcoinTestnet)
+        XCTAssertEqual(address3, "tb1qwgpxgwn33z3ke9s7q65l976pseh4edrzfmyvl0")
```
Comment (Milerius) [inline swift/Tests/HDWalletTests.swift:126]: thanks for adding this

File: src/Binance/Entry.cpp [hunk 2]
```diff
@@ -17,6 +17,13 @@ bool Entry::validateAddress([[maybe_unused]] TWCoinType coin, const std::string&
 }
 
 std::string Entry::deriveAddress([[maybe_unused]] TWCoinType coin, const PublicKey& publicKey, [[maybe_unused]] TWDerivation derivation, [[maybe_unused]] const PrefixVariant& addressPrefix) const {
+    if (std::holds_alternative<Bech32Prefix>(addressPrefix)) {
+        const std::string hrp = std::get<Bech32Prefix>(addressPrefix);
+        if (!hrp.empty()) {
```
Comment (Milerius) [inline src/Binance/Entry.cpp:22]: here u can use auto initializer:

```cpp
if (const auto hrp = std::get<Bech32Prefix>(addressPrefix); hrp) {
}
```

**Review by Milerius — COMMENTED** [review]
Looks ok to me, it's using the same approach as cosmos, @hewigovens can u also take a look?

File: src/Binance/Entry.cpp (hunk 2 above)
Comment (andreibancioiu) [reply src/Binance/Entry.cpp]: Thank you, fixed.

**Review by Milerius — APPROVED** [review]

**Review by Milerius — CHANGES_REQUESTED** [review]
CI is failing

**Comment by andreibancioiu** [comment]
Fixed the failing CI.

**Review by satoshiotomakan — APPROVED** [review]

**Comment by hewigovens** [comment]
We can start adding testnet support, to avoid breaking changes on mobile app side (sending to testnet address will cause losing fund because app doesn't support testnet yet), can we add another method or parameter to validate testnet address?

**Comment by Milerius** [comment]
> We can start adding testnet support, to avoid breaking changes on mobile app side (sending to testnet address will cause losing fund because app doesn't support testnet yet), can we add another method or parameter to validate testnet address?

@andreibancioiu Can you please add unit tests that validate address return false for tbnb address if the HRP is not specified?

File: tests/common/CoinAddressValidationTests.cpp [hunk 3]
```diff
@@ -45,7 +45,9 @@ TEST(Coin, validateAddressBitcoin) {
 
 TEST(Coin, ValidateAddressBinance) {
     EXPECT_TRUE(validateAddress(TWCoinTypeBinance, "bnb12vtaxl9952zm6rwf7v8jerq74pvaf77fcmvzhw"));
-    EXPECT_FALSE(validateAddress(TWCoinTypeBinance, "tbnb12vtaxl9952zm6rwf7v8jerq74pvaf77fkw9xhl"));
+    EXPECT_TRUE(validateAddress(TWCoinTypeBinance, "tbnb1devga6q804tx9fqrnx0vtu5r36kxgp9t4ruzk2"));
```
Comment (Milerius) [inline tests/common/CoinAddressValidationTests.cpp:48]: Should return false if the HRP is not specified.

**Review by Milerius — CHANGES_REQUESTED** [review]

File: tests/common/CoinAddressValidationTests.cpp (hunk 3 above)
Comment (andreibancioiu) [reply tests/common/CoinAddressValidationTests.cpp]: Fixed, and added an extra test point.

**Review by Milerius — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Pull request merged**